### PR TITLE
Create fazeID.md

### PR DESCRIPTION
**Pull request by jettblu**
#### Project Abstract

Faze ID is bringing privacy-preserving face recognition to the browser. The primary applications of Faze ID are Sybil resistance and authentication. The grant proposal encompasses research, model evaluation, and deployment.

#### Grant [level](https://github.com/w3f/Grants-Program#level_slider-levels)

- [x] **Level 1**:  Up to $10,000, 2 approvals
- [ ] **Level 2**:  Up to $30,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 5 approvals (for >$100k: Web3 Foundation Council approval)

#### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied and aptly renamed (`project_name.md`).
- [x] I have read the [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/grant_guidelines_per_category.md).
- [x ] Payment details have been provided (bank details via email _or_ BTC, Ethereum (USDC/DAI) or Polkadot/Kusama (USDT) address in the application).
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process) (see the [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/announcement-guidelines.md)).
- [ ] I prefer the discussion of this application to take place in a private Element/Matrix channel. My username is: `@_______:matrix.org` (change the homeserver if you use a different one)

### CODE REVIEW

**Comment by keeganquigley** [comment]
Thanks for the application @jettblu, I was a fan of your previous Distributed Key Management grant. However, it's not immediately clear to me what Polkadot tech is involved for this project. 

- Are you storing hashes on-chain, or how else are you leveraging substrate tech here?
- Also, I'm curious as to why you are waiting until Milestone 3 to do the research part? Wouldn't it be more helpful to do the research first before building it?
- It sounds like the tech will be fully automated, so what kind of verifications can be done to guard against false positives and that sort of thing?
- Will other wallet projects be able to reuse the product for their wallets?

**Comment by jettblu** [comment]
Hi @keeganquigley! Thanks for the quick response. I've included responses to your concerns below. 

- Overall, this project does not use the Polkadot tech stack but is intended to serve as infrastructure for Polkadot projects. Many wallet and defi projects built on Substrate would benefit from a Sybil resistance solution. That's what we are trying to provide with Faze.
- I believe the paper would benefit from being released as the final milestone. This will allow us to include our implementation in the paper and reference evaluation metrics collected in milestone two. I also expect that our final implementation will diverge from our initial ideas. By doing the paper last, we will have more flexibility with our implementation, hopefully leading to a more useful result.
- False positives are a concern. The evaluation phase will allow us to quantify and improve the error rate. However, we don't expect the model to have 100% accuracy. And in fact, we don't need perfection. Most projects (including our own wallet) would benefit from a probabilistic guarantee that some percentage of users are not bots.  
- Yes! One of our primary goals is to make a Sybil resistance system that can be used across the ecosystem. We will run a hosted version of the backend service (free and open source) and provide instructions for projects to set up their own backend validation system.  As mentioned in the proposal, we would like this to be a plug-and-play system where developers can benefit from Sybil resistance by dropping in a React component and including a few lines of verification logic. 

I know this was a long response. Please let me know if you have any questions. I am open to any feedback or modification requests. For more info, you can also view the [Faze website](https://faze.kryptik.app/).

**Comment by keeganquigley** [comment]
Thanks for the answers @jettblu much appreciated. I will mark the application as ready for review and share it with the committee. Normally we ask for our tech stack to be used somehow but considering the price, it's possible the committee might be willing to make an exception. A couple more questions:
- Who would be the main user base for the product? I don't see that in the application.
- We only fund the tech development itself, so I don't think that paying research participants can be covered by the grants program. Would you be willing to remove this? Also, if you are willing to lower it to a level 1 grant only two approvals will be needed.

**Comment by jettblu** [comment]
The main user base of the platform would be application developers. In the application, I mention an example use case of the Nova wallet airdropping users 'seed money' and using Faze to mitigate bot attacks. I can also imagine Polkadot Defi protocols or social networks using Faze to help with Sybil resistance. 

And yes! I did not realize the research request was out of bounds. I'll update the pr/amounts and denote the grant as tier one. 

Let me know if I've missed anything. Thanks again for helping to clarify the scope.

**Comment by keeganquigley** [comment]
Thanks @jettblu one last question, have you done any temperature checks anywhere to see if this is needed by the community?

**Comment by jettblu** [comment]
Hi, @keeganquigley; thanks for updating the PR. I also updated the application itself to reflect our conversation above. And yes, I have spoken to several developers and teams. For example, at EthDenver in March, a central theme was the lack of decentralized sybil resistance solutions outside Gitcoin Passport. In addition, several CMU alums I have spoken to are also building projects/companies that need a clear way to filter low-quality participants and prevent spam. If successful, Faze will provide this solution (free and open source).

**Comment by randombishop** [comment]
Hi @jettblu !

Thank you for this super interesting grant application.

Will the AI/ML model(s) used in the frontend be released as well in an open source repo?

And will the model building pipeline and code also be released in open source?

Also, can you elaborate more on how the "hashing" works? Is it some kind of internal model embedding?

If it's the latter, will it be possible for someone with access to the model to get an approximation of the face image using the hash?

I am super curious about this use case, can you please share links and infos to educate me as much as possible?

Thank you!

**Comment by jettblu** [comment]
Hi @randombishop !

The model pipeline and fitted weights will all be open-source. You can view some initial modeling [here](https://github.com/jettblu/fazeModel). Of note, the first 20 or so principal components account for over 90% of the variance in facial landmarks. This metric was obtained by running tests on the 'Faces in the Wild' dataset. However, PCA alone is not sufficient for production recognition. Our initial testing showed too many false positives. The final model will likely be a custom NN inspired by industry standards like [Facenet](https://arxiv.org/pdf/1503.03832.pdf). 

Locality-sensitive hashing (LSH) tries to maximize hash collisions for similar values. I've created a simple library for core methods like minhash, which can be found [here](https://github.com/jettblu/hyperFuzz). In general, LSH allows you to compute the similarity between two pieces of data efficiently. In addition, LSH minimizes the amount of data that must be transferred from the client and stored on the authentication server. Minhash is insufficient for our recognition task as we deal with non-overlapping face landmark values. I am currently exploring alternatives like [neighbor-sensitive hashing](https://dl.acm.org/doi/10.14778/2850583.2850589). 

I am still exploring the privacy implications of face encodings. The exact privacy details will likely depend on the encoding method. However, all encoding computation is done on the client, so the server can never access the original image/video feed. 

Most of the work/research I mentioned above has been done in the last few weeks, so there is still a lot to explore and refine. I'm happy to provide more details, but I hope this clarifies some of your questions.

**Review by semuelle — CHANGES_REQUESTED** [review]
Thanks for the application, @jettblu. I agree that sybil resistance is a worthwhile problem to tackle. However, I am very skeptical about [putting facial scans on a public blockchain](https://twitter.com/Snowden/status/1451990496537088000), even if they are hashed.

- With the hash and hashing algorithm being public, you cannot prevent others from brute-forcing it. That would make this a fun experiment, but nothing to safeguard any value with.
- According to Apple, there is a 1 in 1,000,000 chance of someone's face matching your scan. Since fazeID is basically a low-tech version of Face ID, I expect a collision - malicious or not - to be much, much higher.
- I think it's great that you are planning experiments to evaluate the technology. However, I think there is plenty of research (and working products) on video identification from the past 20 years, and it's not sufficiently related to the Substrate/Polkadot ecosystem.

**Comment by randombishop** [comment]
Hi @jettblu !

My opinion is that this is a great line of research worth pursuing; but the core part of it is the model encoding the face information; the guarantees of privacy, deep fake and collision resistance, etc. And these are not directly related to the Polkadot ecosystem.

Once you have a POC solving all these problems, I guess integrating the technology into a Polkadot wallet, building a parachain to manage the data, or other decentralization approach could make sense to be funded by W3F. 

But in its current state, it feels like a ML/AI R&D project, and I personally won't approve it.

**Comment by jettblu** [comment]
Hi @semuelle and @randombishop; I appreciate your skepticism—however, a few clarifying points. 

- **No encodings are being stored on the blockchain.** The first implementation will run on a private authentication server. We will provide a public endpoint, but projects can also set up their own backend. 
- **The primary application is sybil resistance.** An example I cited in the grant is protecting airdrops. In this case, Faze would provide a probabilistic guarantee that, say, 85% of the claimers are unique humans. This is an incredibly valuable use case for decentralized ecosystems like Polkadot, where there are few ways to protect against spam/bots.
- **There will be integrations.** AI is the tool, but again, we are focused on tangible impact within the ecosystem. By the end of the grant period, we will integrate Faze within the open-source Kryptik wallet (to verify initial airdrops to users). We will also release a _plug-and-play React component that can be used by other wallets/defi projects through the Polkadot/Kusama space._  Note that while there has been substantial research on modeling techniques, there is not yet a single implementation available for web3/browser-based applications. 

This is a research project, and, as such, there is no guarantee of success. However, I disagree with the claim that R&D on privacy & model characteristics has little relation to the ecosystem. A strong sybil resistance scheme relies on these properties. In addition to providing the authentication component itself, we need to provide performance metrics that can only be obtained by R&D. 

Please let me know if there are any further modifications I can make or use cases you're interested in. I believe strongly in providing a privacy-first sybil resistance scheme, and I think Faze would be a big step in that direction.

**Comment by randombishop** [comment]
You make good points @jettblu ; but just to clarify what I meant by "not directly related to the Polkadot ecosystem" because I probably didn't phrase it correctly.

I agree that it's a great innovation for the blockchain space in general, but I meant that it also works for Ethereum or any other blockchain; and solves sybil resistance in general, not specifically for Polkadot ecosystem. 

The frontend and models can be re-used across all blockchains, while our program focuses on contributions to the polkadot/kusama ecosystem and codebase in particular.

**Comment by jettblu** [comment]
@randombishop You and the reviewers know the mission of the Web3 Foundation better than anyone else. However, I would say that the extensibility of Faze or any other cross-chain solution is a plus, not a minus. 

The primary requirement is that grants interoperate with the Polkadot/Kusama ecosystem. Faze does this. And as the proverbial 'sprinkles on the top,' it also can integrate with any other protocol. 

The bottom line is that Faze seeks to solve a fundamental problem (Sybil resistance) in the Polkadot ecosystem. 

Just to be clear, I am not trying to quibble over the purpose of the grants program; I am just trying to clarify the 'value add' that Faze would provide.

**Review by takahser — CHANGES_REQUESTED** [review]
@jettblu Thank you for submitting your grant proposal.

I have a few concerns that I'd like to address:

**1) Privacy**
It appears that you're planning to implement client-side feature calculation and hashing, as well as locality-sensitive hashing for matching, which are essential measures for preserving user privacy. However, I would still like to learn more about the robustness of these privacy-preserving techniques and how they can ensure user privacy in various scenarios. Additionally, I'm interested in whether you've considered the compliance aspects that a solution like this would entail.

**2) Security**
One potential attack vector that concerns me is camera spoofing. An attacker could manipulate the video stream by using a virtual camera or intercepting and modifying the stream before it reaches the application, possibly bypassing the biometric authentication. Since the code is not running in a trusted execution environment, it might not be guaranteed that the program sees the actual user.

**3) Match with our program**
It seems like the application is not very specific to Substrate and web3, focusing heavily on AI instead. The deliverables are mostly npm- and react-specific, aside from the research-based deliverables. I'm unsure if our program is the best match for your project.

Please feel free to elaborate on and address my points.

File: applications/fazeID.md [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+- **Costs:** 2,000 USD
+
+
+| Number | Deliverable | Specification |
+| -----: | ----------- | ------------- |
+| **0a.** | License | Apache 2.0 |
+| **0b.** | Documentation | We will provide **inline documentation** of the code.|
+| **0c.** | Testing and Testing Guide | Core functions will be fully covered by comprehensive unit tests, using the Vitest framework.|
+| **0d.** | Docker | We will provide a Dockerfile(s) that can be used to test all the functionality delivered with this milestone. |
+| 1. | Liveness detection Module | This module will issue challenges to the client and ensure the user is attentive and human. Initial ideas involve task and response challenges, such as blinking at requested intervals.|
```
[the remark this answers is not in the document]
Comment (jettblu) [reply applications/fazeID.md:129]: The challenge-response is our primary way to screen deep fakes. This task would be included in the liveness detection module. Challenge-response will also help protect against camera spoofing and ensure users are real and attentive.

File: applications/fazeID.md [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+
+| Number | Deliverable | Specification |
+| -----: | ----------- | ------------- |
+| **0a.** | License | Apache 2.0 |
+| **0b.** | Documentation | We will provide **inline documentation** of the code.|
+| **0c.** | Testing and Testing Guide | Core functions will be fully covered by comprehensive unit tests, using the Vitest framework.|
+| **0d.** | Docker | We will provide a Dockerfile(s) that can be used to test all the functionality delivered with this milestone. |
+| **0e.** | Article| We will publish an article that provides a step by step explanation of the Faze ID system and how each function is implemented in code. This will be published as a blog post. |
+| 1. | Formal Review | This period will be spent reviewing our evaluation results and connecting our results with existing literature. Specific questions we want to address are as follows? What is the model'smperformance? What are the privacy guaruntees of Faze Id? Does hash function selection affect model performance? Is the system viable for production, in terms of complexity and performance?|
```
[the remark this answers is not in the document]
Comment (jettblu) [reply applications/fazeID.md:147]: We have a general idea of the privacy guarantees. However, the details will depend on the specific implementation and design decisions made during development. For instance, we may opt to introduce noise to the images as has been suggested by a few prominent papers, which would affect both encodings and performance.

**Comment by jettblu** [comment]
@takahser, thanks for sharing your concerns. I've addressed your points below. 

**1) Privacy** 

Correct. We want to use client-side feature calculation and locality-sensitive-hashing to help protect user privacy. As mentioned in one of the comments above, the full privacy implications of this approach will depend on the precise implementation. However, we _can_ guarantee that no raw video or images will be sent to the server for processing. This is already a big step for privacy. As for compliance, our solution will be open source. That means anyone can run a hosted version of the auth server. While that doesn't remove all compliance considerations, it provides more flexibility for each provider to set up policies. 

**2) Security**

 Both camera spoofing and the lack of a trusted execution environment are part of the challenge. Liveness detection will help with the spoofing vector. In addition, we are developing Faze primarily as a Sybil-resistance tool. The system will not be perfect, but if we can guarantee that 80% of users who pass validation are human, that is a big win for apps that struggle with spam and bots.

**3) Program Fit** 

You are correcting that the primary technical challenge is related to machine learning (and privacy). In addition, we are committed to building a one-line react-based auth plugin for developers, hence the npm and react-related deliverables. However, Faze is specifically meant to benefit permissionless networks like blockchains. Since most applications and potential users are decentralized protocols, it makes sense (in our mind, at least) to apply for a W3F grant.

If you have any more questions or concerns, please don't hesitate to let me know. Faze would be a valuable tool for many applications, including our own wallet, and we would deeply appreciate support from the foundation.

File: applications/fazeID.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (takahser) [reply applications/fazeID.md:129]: Are you planning to write this logic yourself or are you using an existing solution for that? If so, which one? I'm not an expert in the area but I'm aware of tools like www.synthesia.io that allows you to dynamically create deepfakes. With enough computing power I'd imagine to be able to render these real-time.

File: applications/fazeID.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (takahser) [reply applications/fazeID.md:147]: I see, thanks for clarifying.

File: applications/fazeID.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (jettblu) [reply applications/fazeID.md:129]: Our assumption should be that an attacker would have enough computing power to render real-time deep fakes. Some of the logic we are working on would be custom, but we are also looking to implement recent research on real-time deep fake detection. In January, a team from Berkeley achieved >90% detection accuracy with a new method called [D-Captcha.](https://arxiv.org/abs/2301.03064). We would like to implement and extend core principles from this paper. 

Again, all of this is done with a focus on providing sybil-resistance for permissionless systems.

File: applications/fazeID.md [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+| -----: | ----------- | ------------- |
+| **0a.** | License | Apache 2.0 |
+| **0b.** | Documentation | We will provide **inline documentation** of the code.|
+| **0c.** | Testing and Testing Guide | Core functions will be fully covered by comprehensive unit tests, using the Vitest framework.|
+| **0d.** | Docker | We will provide a Dockerfile(s) that can be used to test all the functionality delivered with this milestone. |
+| 1. | Liveness detection Module | This module will issue challenges to the client and ensure the user is attentive and human. Initial ideas involve task and response challenges, such as blinking at requested intervals.|
+| 2. | Model Evaluation | We will test our combined recognition and liveness model on at least 50 paid research participants. Desired metrics include model recall, false positive rate, and attention tracking.|
+| 3. | Results Summary| We will publish our evaluation results and insights as a long form article. These metrics will serve as reference for future integrations (for instance, defi aps may use false positive rates to quantify risk scores).|
+
+### Milestone 3 — Rsearch Paper and Faze Package
```
Comment (takahser) [inline applications/fazeID.md:133]:
```suggestion
### Milestone 3 — Research Paper and Faze Package
```

File: applications/fazeID.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (takahser) [reply applications/fazeID.md:129]: > In January, a team from Berkeley achieved >90% detection accuracy with a new method called [D-Captcha.](https://arxiv.org/abs/2301.03064). We would like to implement and extend core principles from this paper.

Could you integrate this into the doc?

**Review by takahser — CHANGES_REQUESTED** [review]
@jettblu thanks for answering the questions. Currently, I cannot approve this due to its lack of relevance to the web3 and Substrate ecosystems. If you can incorporate integration with a Substrate-based solution and demonstrate a compelling use case where your solution is beneficial, I would be open to reconsidering my decision.

**Comment by jettblu** [comment]
Hey @keeganquigley, I reviewed the RFP list and found a [project](https://github.com/w3f/Grants-Program/blob/master/docs/RFPs/Open/grant_management_webapp.md) (proposed. by @randombishop) I'm excited about. However, I have a few questions. What do you think is the best way to communicate on open RFPs? Thanks for your help!

**Comment by keeganquigley** [comment]
Hi @jettblu glad to hear you are interested in this RFP! Email works fine at grants@web3.foundation or we can start a private chat room in Element as well, either works! Just let me know.

**Comment by jettblu** [comment]
Ok, thanks for letting me know. I prefer to use Element. My username is 'jettblu'. I have already joined the W3F Grants Community room.
